### ISSUE TITLE
rfc: better native bindings

### ISSUE THREAD

**Issue by thehowl** [gnolang/gno#814]
This is an issue to propose and discuss a better integration and implementation of native Go bindings in the Gno standard libraries, trying to remove places for pitfalls and for erroneous behaviour while making it easier to contribute and expand functionality in the standard libraries.

## Background

_See also: #808, #668, #522_

The standard library is seemingly designed to allow for running Gno code both in a context where stdlibs are loaded as .gno files from the `gnovm/stdlibs` directory, as native bindings (mostly useful for the filetests in `gnovm/tests`), and also where code is precompiled. This is a bit disorganised in different places, and has generated the issues mentioned above.

- `stdlibWhitelist` in gnovm/pkg/gnolang/precompile.go has to be kept in sync with the libraries  from the Go stdlib which we have implemented. Although furthermore we also have to specify custom rules on how to handle the "std" package, and crucially would currently require more manual intervention if we wanted to precompile ie packages importing `crypto/chacha20` (which is a custom Gno library).
- Many functions implemented in `stdlibs/stdlibs.go` are defined using `pn.DefineNative` when this is not really required because they don't intrinsically need to know about the context, but they do something which would be hard/slow to do in pure gno, with such an example being the `sha256.Sum256` function.
- The test environment exposes more functions in a manner that is not really transparent - these are defined in `gnovm/tests/imports.go`, including packages like os, fmt, base64, etc. While it largely makes sense for these not to be exposed in standard gnovm contexts, it should be (1) be more visible to tooling (like `gno doc`) that these functions exist and that (2) they are available when writing tests.

## Proposal

I propose to remove native bindings the way they are done right now, and instead use a similar mechanism's to Go's with assembly code: if a function only has a declaration without a body, then it is expected to be implemented in Go. Because of Gno's interpreted nature, this is only expected and supported in the standard library, where we can use tooling to do static code analysis and pregenerate all bindings to native Go functions.

Additionally, this respects the `_test.gno` suffix - if a function is declared in a file with such a suffix, it will only be available in test and filetest contexts.

Functions in Go standard code may be implemented directly as functions with the same signature as in the Gno code, or also in `func Init(store gno.Store, pn *gno.PackageNode)` functions, which work the same way as `stdlibs.InjectPackage` (except PkgPath is guaranteed to be the same as the path the file is placed in, relative to the `stdlibs` directory). 

From a precompiler perspective, as most functions' native bindings will be something like

```go
// json.gno
func MarshalJSON(v interface{}) ([]byte, error)

// json.go
import "encoding/json"

func MarshalJSON(v interface{}) ([]byte, error) {
    return json.MarshalJSON(v)
}
```

we can adopt a simple philosophy of "inlining" function calls similar to how they're done with go - if they can be simply replaced with another statement, then it will be done that way. Otherwise, if the native functions are more complex, then the package "$gnoimportpath/gnovm/stdlibs/encoding/json" (for instance) is imported and the native function called.

If a function is implemented in the `Init` function as described above, the definition in Init will take precedence. However an additional function can be defined in Go to specify the behaviour of the function when precompiled to native Go - this way, functions like `std.IsOriginCall` can be implemented using the machine context in Gno, but return a panic in the precompiled version.

std.Address, std.Coins and std.Coins, which are types mapped to internal go types, could be specified using a directive like `//gno:link Coins "github.com/gnolang/gno/tm2/pkg/std".Coins`, similar to `go:linkname`.

Having functions declared in .gno files and implemented in .go files alongside further allows tools like `gno doc` to work by only doing static analysis on gno stdlib files, without having to add specific exceptions to support the standard library.

The important part is that all of this is done using generated code which performs analysis on the stdlibs directory to then ensure that we have a central package where to choose between the different contexts and which is also able to give some metadata for precompilation.

**Comment by ajnavarro** [comment]
I think that maybe we don't need any precompile step to do, and we can just execute native code when interpreting the code on the VM if we detect that the function called is from the standard library. We will need to give to these functions a specific gas fee too.

**Comment by thehowl** [comment]
@ajnavarro we don't do precompile steps in normal executions, but precompilation is a feature supported by cmd/gno and I think we should continue to support.

And yes, indeed while we still need to sort out gas fees for native function it currently works like you said. Except that now everything's in a big file containing all the native functions (mostly stdlibs/stdlibs.go)

**Comment by tbruyelle** [comment]
> One concern is that precompilation (I explained better below how that works) would probably have more trouble if they're implemented this way.

Yes probably, let's do it with function without body + `Init` first. Then we'll see if we can simplify with a only function without body.

Thx for the explanation.

**Comment by moul** [comment]
Related with #239 and #498

**Comment by thehowl** [comment]
I've incorporated the latest ideas in the comments into the OP, and also added a lengthier example section showing exactly how I envision both stdlib code and realm code (and precompiled).

Feedback is very welcome. I'll start working on a PoC soon.

**Comment by ajnavarro** [comment]
Leaving a different approach here, just to continue the discussion and checking pros and cons:

using the following example:

#### Example user of `std` package, and generated precompiled code

```go
// examples/gno.land/r/r1/r1.gno
package r1

import "std"

func SampleFunction() {
	std.EncodeBech32("r1", [20]byte{})
	std.Hash([]byte(std.GetOrigCaller())
}
```

```go
// examples/gno.land/r/r1/r1.gno.gen.go (ie. precompiled)
package r1

import (
	"github.com/gnolang/gno/gnovm/stdlibs/std"
)

func SampleFunction() {
	std.EncodeBech32("r1", [20]byte{})
	std.Hash([]byte(std.GetOrigCaller())
}
```

- `std` package will be implemented in go, so we can use it as a dependency for go precompiled code, and internally on the VM.
- There will be no difference between the go and gno code, so the only thing that will change is the import.
- When code is executed inside the VM, we can use reflection to call functions on specific packages, `std` in this case.
- No need to keep track of a huge list of methods on the VM, any change to `std` package will be reflected on VM and generated go code.

WDYT?

**Comment by thehowl** [comment]
@ajnavarro 

One question, from your example: how does `std.GetOrigCaller()` gather access to `*gno.Machine`? Especially for the `std` package, many function calls require access to the VM's data. Otherwise, the proposals look very similar

**Comment by ajnavarro** [comment]
We could call something like `std.With(vm).GetOrigCaller()` from the VM to make it available. When the code is executed from Go, we can leave a default mocked VM as the default or use any other implementation. But these might be ideas for the future. `std.GetOrigCaller()` will call under the hood to `std.With(&MockVM{}).GetOrigCaller()`  (or a nil VM) when called from Go, and that call to GetOrigCalled will be translated to `std.With(vm).GetOrigCaller()` when called from Gno VM using reflection.

What I'm trying to do here is to avoid having code generation, and make everything as straightforward as possible to allow future IDEs to autocomplete and things like that.

Feel free to discard the idea if you feel it doesn't fit well (you have a bigger context here than me). I'm just hoping that this proposal helps to validate the idea.
